File: archive/2020/solutions/Icebergs.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;


struct Ship
{
    bool isDestroyed;
    long long x, y;
};

int numShips;
vector<Ship> allShips;
vector<int> shipDestroyIndices;
vector<vector<int>> closest;

Ship curr;

bool shipCmp(int s1, int s2){
    long long distSqr1 = (allShips[s1].x - curr.x) * (allShips[s1].x - curr.x) + (allShips[s1].y - curr.y) * (allShips[s1].y - curr.y);
    long long distSqr2 = (allShips[s2].x - curr.x) * (allShips[s2].x - curr.x) + (allShips[s2].y - curr.y) * (allShips[s2].y - curr.y);
    if(distSqr1 == distSqr2){
        return s1 < s2;
    }
    return distSqr1 < distSqr2;
}

int main(){
    cin >> numShips;
    allShips.resize(numShips);
    shipDestroyIndices.resize(numShips);
    closest.resize(numShips);

    for(int i = 0; i < numShips; i++){
        int shipX, shipY;
        cin >> shipX >> shipY;
        allShips[i] = Ship{false, shipX, shipY};
        shipDestroyIndices[i] = 0;
        for(int j = 0; j < numShips; j++){
            if(j == i){
                continue;
            }
            closest[i].push_back(j);
        }
    }
    
    for(int i = 0; i < numShips; i++){
        curr = allShips[i];
        sort(closest[i].begin(), closest[i].end(), shipCmp);
    }

    int output = 0;
    int numDestroyed = 0;
    while(numDestroyed < numShips - 1){
        for(int i = 0; i < numShips; i++){
            if(allShips[i].isDestroyed){
                continue;
            }
            numDestroyed++;
            curr = allShips[i];
            output = i;
            for(int j = shipDestroyIndices[i]; j < numShips - 1; j++){
                if(!allShips[closest[i][j]].isDestroyed){
                    allShips[closest[i][j]].isDestroyed = true;
                    shipDestroyIndices[i]++;
                    break;
                }
            }
        }
    }

    cout << output+1 << '\n';
}

File: archive/2020/solutions/Icebergs.py
curr_ship = [0, 0, 0]

def get_dist(val):
    return (all_ships[val][0] - curr_ship[0]) * (all_ships[val][0] - curr_ship[0]) + (all_ships[val][1] - curr_ship[1]) * (all_ships[val][1] - curr_ship[1])


num_ships = int(input())

all_ships = [[0, 0, 0] for i in range(num_ships)]
curr_destroyed = [1] * num_ships
closest = [[j for j in range(num_ships)] for i in range(num_ships)]

for i in range(num_ships):
    in_line = input().split()
    all_ships[i][0] = int(in_line[0])
    all_ships[i][1] = int(in_line[1])
    
for i in range(num_ships):
    curr_ship = all_ships[i]
    closest[i].sort(key=get_dist)

num_destroyed = 0
output = 0
while num_destroyed < num_ships - 1:
    for i in range(num_ships):
        if all_ships[i][2]:
            continue
        num_destroyed += 1

        curr_ship = all_ships[i]
        output = i
        for j in range(curr_destroyed[i], num_ships):
            if not all_ships[closest[i][j]][2]:
                all_ships[closest[i][j]][2] = 1
                curr_destroyed[i] += 1
                break

print(output+1)
